### ISSUE TITLE
Configuring Data Protection in Blob storage with encryption fails

### ISSUE THREAD

**Issue by span** [aspnet/AspNetCore#10299]
Using dot net core 2.2.

The initialisation of data protection and encryption of key files in Azure Blob Storage is not behaving as expected.

I have created a proof of concept project with logs and more details here:
https://github.com/span/azure-dp-conf

## Setup

I expect the data protection service to be configured like this.

```c#
services
    .AddDataProtection(options =>
    {
        options.ApplicationDiscriminator = 
            Configuration["DataProtectionKeys:applicationDiscriminator"];
    })
    .PersistKeysToAzureBlobStorage(blobContainer, "azuredpconf-keys.xml")
    .ProtectKeysWithAzureKeyVault(
        Configuration["AzureKeyVault:DataProtectionKey"],
        Configuration["AzureKeyVault:ClientId"],
        Configuration["AzureKeyVault:ClientSecret"]);
```

This fails however since no key file is generated in 
`PersistKeysToAzureBlobStorage` and `ProtectKeysWithAzureKeyVault` seems
to require the key file.

A log of the failing run can be found in `./logs/First-RunFails.log`.

### Expected result

Host launches with no errors and key file is generated in the blob storage.

### Actual result

Host throws an exception and no key file has been generated in the blob storage.

<details>
<summary>Stack trace from First-Run-Fails.txt</summary>
<pre>
fail: Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider[48]
      An error occurred while reading the key ring.
System.UriFormatException: Invalid URI: The format of the URI could not be determined.
   at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
   at System.Uri..ctor(String uriString, UriKind uriKind)
   at Microsoft.Azure.KeyVault.ObjectIdentifier..ctor(String collection, String identifier)
   at Microsoft.Azure.KeyVault.KeyVaultClientExtensions.WrapKeyAsync(IKeyVaultClient operations, String keyIdentifier, String algorithm, Byte[] key, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.DataProtection.AzureKeyVault.AzureKeyVaultXmlEncryptor.EncryptAsync(XElement plaintextElement)
   at Microsoft.AspNetCore.DataProtection.AzureKeyVault.AzureKeyVaultXmlEncryptor.Encrypt(XElement plaintextElement)
   at Microsoft.AspNetCore.DataProtection.XmlEncryption.XmlEncryptionExtensions.EncryptIfNecessary(IXmlEncryptor encryptor, XElement element)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager.Microsoft.AspNetCore.DataProtection.KeyManagement.Internal.IInternalXmlKeyManager.CreateNewKey(Guid keyId, DateTimeOffset creationDate, DateTimeOffset activationDate, DateTimeOffset expirationDate)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager.CreateNewKey(DateTimeOffset activationDate, DateTimeOffset expirationDate)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.CreateCacheableKeyRingCore(DateTimeOffset now, IKey keyJustAdded)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.Microsoft.AspNetCore.DataProtection.KeyManagement.Internal.ICacheableKeyRingProvider.GetCacheableKeyRing(DateTimeOffset now)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.GetCurrentKeyRingCore(DateTime utcNow)
info: Microsoft.AspNetCore.DataProtection.Internal.DataProtectionStartupFilter[0]
      Key ring failed to load during application startup.
System.UriFormatException: Invalid URI: The format of the URI could not be determined.
   at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
   at System.Uri..ctor(String uriString, UriKind uriKind)
   at Microsoft.Azure.KeyVault.ObjectIdentifier..ctor(String collection, String identifier)
   at Microsoft.Azure.KeyVault.KeyVaultClientExtensions.WrapKeyAsync(IKeyVaultClient operations, String keyIdentifier, String algorithm, Byte[] key, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.DataProtection.AzureKeyVault.AzureKeyVaultXmlEncryptor.EncryptAsync(XElement plaintextElement)
   at Microsoft.AspNetCore.DataProtection.AzureKeyVault.AzureKeyVaultXmlEncryptor.Encrypt(XElement plaintextElement)
   at Microsoft.AspNetCore.DataProtection.XmlEncryption.XmlEncryptionExtensions.EncryptIfNecessary(IXmlEncryptor encryptor, XElement element)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager.Microsoft.AspNetCore.DataProtection.KeyManagement.Internal.IInternalXmlKeyManager.CreateNewKey(Guid keyId, DateTimeOffset creationDate, DateTimeOffset activationDate, DateTimeOffset expirationDate)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager.CreateNewKey(DateTimeOffset activationDate, DateTimeOffset expirationDate)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.CreateCacheableKeyRingCore(DateTimeOffset now, IKey keyJustAdded)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.Microsoft.AspNetCore.DataProtection.KeyManagement.Internal.ICacheableKeyRingProvider.GetCacheableKeyRing(DateTimeOffset now)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.GetCurrentKeyRingCore(DateTime utcNow)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.GetCurrentKeyRing()
   at Microsoft.AspNetCore.DataProtection.Internal.DataProtectionStartupFilter.Configure(Action`1 next)
</pre>
</details>

## Workaround

I have found two workarounds. 

1. Create the key file manually in blob storage before launching the Host.
2. Remove the `ProtectKeysWithAzureKeyVault` during first run to make sure 
the key file is created and then re-enable the key encryption.

**Comment by blowdart** [comment]
@natemcmaster could you take a look please?

**Comment by natemcmaster** [comment]
ACK. Will take a look in the next milestone.

**Comment by Pilchie** [comment]
@anurse do you know who could take a look at this?

**Comment by anurse** [comment]
@HaoK has done some work here before, and I could find someone, but we're pretty fully booked for preview 8 so it would likely increase risk in other areas. There is a workaround posted, which doesn't at all remove the need to fix this but perhaps indicates that it could be a candidate for 3.1 (which isn't far after 3.0 anyway)

**Comment by Pilchie** [comment]
@HaoK can you take a look please?

**Comment by blowdart** [comment]
This is a candidate for 3.1, due to the presence of a work around.

@anurse @Pilchie can we get a 3.1 milestone yet?

**Comment by Pilchie** [comment]
There is one.

**Comment by HaoK** [comment]
Linking to a related issue https://github.com/aspnet/AspNetCore.Docs/issues/12592

**Comment by HaoK** [comment]
@blowdart after looking at this a bit, given that there is no `PersistKeysToAzureKeyValue` unlike `PersistKeysToDbContext` or `PersistKeysToAzureBlobStorage`, I believe the current behavior is probably intentional, in that you have to manually persist the key to azure key value, and then it will work fine

**Comment by HaoK** [comment]
That said I could be confused, so could you take another look @blowdart :)

**Comment by blowdart** [comment]
That seems odd, but also kind of makes sense, as you want to limit privileges to keyvault. Is it documented anywhere though?

**Comment by span** [comment]
Interesting. From [the docs](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.dataprotection.azuredataprotectionbuilderextensions.persistkeystoazureblobstorage?view=aspnetcore-3.0)

> Configures the data protection system to persist keys to the specified path in Azure Blob Storage.

I would expect the data protection system to persist the keys.

Looking at the remarks they say:

> The container referenced by container must already exist.

So nothing about creating the file manually, only the container.

**Comment by HaoK** [comment]
So basically it seems like ProtectWithAzureKeyVault doesn't work unless the key already exists.  So the workaround where you run it once without Protect to create the keys, and then add it back, might be the expected flow...  Sound reasonable @blowdart ?  We can certainly update the docs to call this quirk out?

**Comment by blowdart** [comment]
Oh wait, the key part in keyvault is the wrapping key. @span data protection keys aren't stored in keyvault.

Ok this makes sense now I remember everything. Yes, a key to encrypt the other keys needs to exist first, just like an X509 certificate needs to exist if you use that to protect keys. This seems like a doc failure.

**Comment by span** [comment]
Yes.

I always thought that `PersistKeysToAzureBlobStorage` would return _after_ the file had been created (if it didn't exist) and then `ProtectKeysWithAzureKeyVault` would encrypt the keys in blob storage with some key in keyvault.

If it is a doc failure perhaps these methods should not be allowed to be chained either?

**Comment by blowdart** [comment]
Oh wait, it's the file creation that fails? That does feel a little weird. The file should get created when a key is first generated. OK this is odd. @HaoK we can chat about this at triage today.

**Comment by span** [comment]
I did a migration of the [proof of concept](https://github.com/span/azure-dp-conf/tree/netcoreapp3.0) project today and it seems to be fixed in 3.0. No idea why but it seems the data protection changes sorted everything out.

**Comment by HaoK** [comment]
Great, glad its already fixed in 3.0 @span
